controller_manager:
  ros__parameters:
    update_rate: 1000  # Hz

    joint_state_broadcaster:
      type: joint_state_broadcaster/JointStateBroadcaster
      ros__parameters:
        use_local_topics: true

    imu_sensor_controller:
      type: imu_sensor_broadcaster/IMUSensorBroadcaster

    quadruped_broadcaster:
      type: quadruped_mpc/quadruped_broadcaster

    state_estimator:
      type: quadruped_mpc/state_estimator

    balance_controller:
      type: quadruped_mpc/balance_controller

# Individual controller configurations
joint_state_broadcaster:
  ros__parameters: {}  # Empty is fine, it will auto-discover joints

imu_sensor_controller:
  ros__parameters:
    sensor_name: imu_sensor
    frame_id: imu_link

quadruped_broadcaster:
  ros__parameters:
    # doesn't use any of the joints, but they are required for the controller_manager
      joints:
        - fr_hip
        - fr_knee
        - fl_hip
        - fl_knee
        - rr_hip
        - rr_knee
        - rl_hip
        - rl_knee
      imu_name: imu_sensor

      # Hardware params
      gravity: 9.81
      hardware_height: 0.25
      hardware_width: 0.25
      hardware_length: 0.25

      # Balance controller params
      balance_m: 25.0
      balance_kp_pos: 500.0
      balance_kd_pos: 5.0
      balance_kp_rot: 500.0
      balance_kd_rot: 5.0
      p_c_des: [0.0, 0.0, 0.2]
      v_c_des: [0.0, 0.0, 0.0]
      th_c_des: [1.0, 0.0, 0.0, 0.0]
      om_c_des: [0.0, 0.0, 0.0, 1.0]
      S: [1.0, 0.0, 0.0, 0.0, 0.0, 0.0,
          0.0, 1.0, 0.0, 0.0, 0.0, 0.0,
          0.0, 0.0, 1.0, 0.0, 0.0, 0.0,
          0.0, 0.0, 0.0, 1.0, 0.0, 0.0,
          0.0, 0.0, 0.0, 0.0, 1.0, 0.0,
          0.0, 0.0, 0.0, 0.0, 0.0, 1.0]
      alpha: 1.0
      beta: 1.0
      C_row_support: [1.0, 1.0, 1.0]
      C_row_swing: [1.0, 1.0, 1.0]
      d_row_support: [1.0, 1.0, 1.0]
      d_row_swing: [1.0, 1.0, 1.0]

state_estimator:
  ros__parameters:
    joints:
      - fr_hip
      - fr_knee
      - fl_hip
      - fl_knee
      - rr_hip
      - rr_knee
      - rl_hip
      - rl_knee
    state_interfaces:
      - position
      - velocity
      - effort

balance_controller:
  ros__parameters:
    joints:
      - fr_hip
      - fr_knee
      - fl_hip
      - fl_knee
      - rr_hip
      - rr_knee
      - rl_hip
      - rl_knee
    command_interfaces:
      - position
      - velocity
      - effort